### PR TITLE
Elastic Training support in DeepSpeed

### PR DESCRIPTION
**Pull request by aj-prime**
This PR adds basic elastic training support to DeepSpeed using Elastic Agent in PyTorch.

Requirement:
PyTorch >= 1.11.x

## Modifications:
1. Added runtime options for elastic training in deepspeed launcher.
2. Implemented a new Elastic training agent for DeepSpeed based on PyTorch Elastic Agent.
3. Requires user to save and load latest checkpoint
4. Added Elasticity V 0.2 that relaxes one batch size constraint.  
5. Tested Megatron and HelloDeepSpeed models on Deepspeed cluster

## Related Repos:

Megatron: https://github.com/microsoft/Megatron-DeepSpeed/tree/arpan/elastic_training
DeepSpeedExamples: https://github.com/microsoft/DeepSpeedExamples-internal/tree/arpan/elastic_scripts

Instructions to launch elastic training  
Megatron Model: https://github.com/microsoft/Megatron-DeepSpeed/blob/arpan/elastic_training/examples/MoE/README_elastic.md
HelloDeepSpeed: https://github.com/microsoft/DeepSpeedExamples-internal/edit/arpan/elastic_scripts/elastic_training/README.md

Log: 
We start training on allocation 1 (2 nodes) and drop one node after 40th iteration. 

|     Iterations    |     World Size    |     Number of Nodes in Training    |                     |
|-------------------|-------------------|------------------------------------|---------------------|
|                   |                   |     Allocation 1                   |     Allocation 2    |
|     0-40          |     8             |     2                              |     0               |
|     40-60         |     4             |     1                              |     0               |
|     60-80         |     12            |     1                              |     2               |
|     80-100        |     8             |     1                              |     1               |
|     100-140       |     12            |     2                              |     1               |
|     140-160       |     16            |     2                              |     2               |
|     160-180       |     12            |     2                              |     1               |
|     180-200       |     8             |     1                              |     1               |
|     200 -         |     4             |     1                              |     0               |

## Workflow 
PR is divided into two modules that can be enabled separately or simultaneously.

Runner module that focuses on relaunching of training in scale-up or scale-down event 
workflow_runner_module png

Valid Batch Size computation in Elastic training 
workflow_elasticity_v2 png

### CODE REVIEW

File: deepspeed/launcher/launch.py [hunk 1]
```diff
@@ -66,6 +67,20 @@ def parse_args():
                         help="Skip prepending the training script with "
                         "'python' - just execute it directly.")
 
+    parser.add_argument("--enable_elastic_training",
+                        action="store_true",
+                        help="Enable elastic training support.")
+
+    parser.add_argument("--min_nodes",
```
Comment (tjruwase) [inline deepspeed/launcher/launch.py:74]: Please rename as `--min_elastic_nodes` and `--max_elastic_nodes` for clarity.

File: deepspeed/launcher/runner.py [hunk 2]
```diff
@@ -77,6 +78,18 @@ def parse_args(args=None):
                         help="Total number of worker nodes to run on, this will use "
                         "the top N hosts from the given hostfile.")
 
+    parser.add_argument("--min_num_nodes",
```
Comment (tjruwase) [inline deepspeed/launcher/runner.py:81]: Please rename as above

Comment (aj-prime) [reply deepspeed/launcher/runner.py]: Thanks, I have renamed them

File: deepspeed/launcher/launch.py [hunk 3]
```diff
@@ -190,13 +246,33 @@ def main():
...
-
-        process = subprocess.Popen(cmd, env=current_env)
-        processes.append(process)
+        cmd_args = cmd[1:]
+
+        rdzv_configs: Dict[str, str] = {'timeout': 100}
+        rdzv_parameters = RendezvousParameters(backend='c10d',
+                                               endpoint=args.master_addr + ":" +
+                                               str(args.master_port),
+                                               run_id='123456789',
```
Comment (jeffra) [inline deepspeed/launcher/launch.py:256]: Discussed offline. Let's fetch `run_id` from an env variable, if it's not set revert back to the default you defined in constants `ELASTIC_TRAINING_ID_DEFAULT`.

Comment (aj-prime) [reply deepspeed/launcher/launch.py]: Thanks for catching it.

**Review by jeffra — APPROVED** [review]

**Pull request merged**